Multiple fixes for Automate and Workflow Builder resources.

ncerny opened the pull request:
Signed-off-by: Nathan Cerny <ncerny@gmail.com>

### Description

There are several new bugs in the chef_stack resources due to changes in chef-client, automate, and delivery-cli.

### Issues Resolved

https://github.com/ncerny/chef_stack/issues/32
https://github.com/ncerny/chef_stack/issues/33
https://github.com/ncerny/chef_stack/issues/34

### Check List

- [ ] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

schisamo commented on resources/automate.rb:
```diff
@@ -106,14 +102,11 @@
-        not_if "delivery-ctl list-enterprises --ssh-pub-key-file=/etc/delivery/builder_key.pub | grep -w #{ent}"
-        only_if 'delivery-ctl status'
-      end
+  (new_resource.enterprise.is_a?(String) ? [new_resource.enterprise] : new_resource.enterprise).each do |ent|
```
You can actually use `Array` here which returns a flattened array in all cases...it is also preferred to call the property directly now (e.g. now `new_resource.`):

```
Array(enterprise).each do |ent|
```

Here is an example of the `Array` method usage:
```irb
➜  irb
Welcome to IRB. You are using ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]. Have fun ;)
>> Array('foo') #=> ["foo"]
>> Array(['foo']) #=> ["foo"]
```

ncerny replied:
> it is also preferred to call the property directly now

I've had mixed results with that (which I assume is my own misunderstanding of scope and context within certain methods), but that would simplify a lot!

wrightp commented on resources/wf_builder.rb:
```diff
@@ -247,12 +260,13 @@
           -e #{new_resource.automate_enterprise} \
           -u #{new_resource.automate_user}").run_command
-        ::File.write(::File.join(home_dir, '.ssh/authorized_keys'), JSON.parse(runner.stdout)['openssh_public_key'])
+        ::File.write(::File.join(home_dir, '.ssh/authorized_keys'), JSON.parse(runner.stdout.gsub(/\e\(B|\e\[m|\e\[37m/, ''))['openssh_public_key'])
```
Use `--no-color` in the command versus stripping out the color syntax from the output.

wrightp replied:
`-non-interactive` may be a better option and hopefully removes the color syntax as well.

wrightp commented:
Let me know if you need assistance resolving conflicts.

ncerny commented on resources/wf_builder.rb:
```diff
@@ -72,19 +72,18 @@
 
   file '/opt/chefdk/embedded/ssl/certs/cacert.pem' do
-    content cacert_pem
+    content @cacert_pem
```
This should probably be lazy { }

ncerny commented:
I have not yet tested the changes made - I'd like to do that before we merge.

I added --non-interactive because it makes sense, but I couldn't find in the code where that actually changes anything, so I also added --no-color.  When I get a test infra stood up, I'll test both commands and see if we can eliminate one.

wrightp commented on resources/wf_builder.rb:
```diff
@@ -69,10 +69,22 @@
 
-  execute 'cat /etc/chef/trusted_certs/*.crt >> /opt/chefdk/embedded/ssl/certs/cacert.pem'
+  file '/opt/chefdk/embedded/ssl/certs/cacert.pem' do
+    content @cacert_pem
```
Instead of lazy we should not use the instance variable syntax since this won't be used as an accessor.

Preferably set `cacert_pem = nil` above the ruby block and reference it by it's name.

wrightp replied:
Or set it like you did in line 77. Either way.

andy-dufour commented:
Just hit a bug that's resolved by this, anything I can do to help get this merged?

andy-dufour commented on resources/wf_builder.rb:
```diff
@@ -247,7 +259,9 @@
 
-        runner = Mixlib::ShellOut.new("delivery api post runners \
+        # TODO: Rework this to call the API directly, so we don't have delivery-cli formatting things.
+        runner = Mixlib::ShellOut.new("delivery --non-interactive --no-color \
```
Order matters here :( `delivery api post runners --non-interactive --no-color` has the intended behavior but `delivery --non-interactive --no-color api post runners` does not, if you order it like this: https://github.com/chef-cookbooks/chef-ingredient/blob/ad/wf-builder-encoding-fix/resources/wf_builder.rb#L263-L264 I see the desired outcome

tduffield replied:
Here is an example of how you can "call the API directly": https://github.com/chef-cookbooks/expeditor/blob/master/libraries/helpers.rb#L40-L48. 

Here is an example of it in use. https://github.com/chef-cookbooks/expeditor/blob/0fd150f127404a56789f96955a8fa279c01a89dc/libraries/helpers.rb#L23

Faraday comes with Chef, and the token data is easy enough to find in various scenarios.

tduffield commented on resources/automate.rb:
```diff
@@ -102,14 +101,11 @@
-        not_if "automate-ctl list-enterprises --ssh-pub-key-file=/etc/delivery/builder_key.pub | grep -w #{ent}"
-        only_if 'automate-ctl status'
-      end
+  Array(enterprise).each do |ent|
```
`automate-ctl` will not allow you to create more than one enterprise on modern versions of Automate.
